AVR GPIO macro defines more readable

mtei opened the pull request:
## Description

 * A little easier to read the definition of the GPIO control macro for AVR.
   quantum/quantum.h, quantum/config_common.h
 * Modified split_common/serial.c to use qmk_firmware standard GPIO control macro.
   quantum/split_common/serial.c

There is no change in compilation results for this PR.

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented on quantum/quantum.h:
```diff
@@ -149,28 +149,26 @@ extern layer_state_t default_layer_state;
-            PIN_ADDRESS(pin, 1) &= ~ _BV(pin & 0xF);\
-            PIN_ADDRESS(pin, 2) |=   _BV(pin & 0xF);\
-            })
+    #define setPinInput(pin) DDRx_ADDRESS(pin) &= ~ _BV(pin & 0xF)
```
@jackhumbert ?

mtei replied:
In this case, I like using the traditional 'do {} while (0)` idiom instead of the gcc extension.

drashna commented:
This looks good to me.  

@pelrun ?

fauxpark commented on quantum/quantum.h:
```diff
@@ -149,28 +149,26 @@ extern layer_state_t default_layer_state;
-            PIN_ADDRESS(pin, 1) &= ~ _BV(pin & 0xF);\
-            PIN_ADDRESS(pin, 2) |=   _BV(pin & 0xF);\
-            })
+    #define setPinInput(pin) DDRx_ADDRESS(pin) &= ~ _BV(pin & 0xF)
```
```suggestion
    #define setPinInput(pin) DDRx_ADDRESS(pin) &= ~_BV(pin & 0xF)
```

mtei replied:
Thanks.

fauxpark commented on quantum/quantum.h:
```diff
@@ -149,28 +149,26 @@ extern layer_state_t default_layer_state;
-            })
+    #define setPinInput(pin) DDRx_ADDRESS(pin) &= ~ _BV(pin & 0xF)
+    #define setPinInputHigh(pin) do {\
+            DDRx_ADDRESS(pin) &= ~ _BV(pin & 0xF);\
```
```suggestion
            DDRx_ADDRESS(pin) &= ~_BV(pin & 0xF);\
```

mtei replied:
Thanks.

fauxpark commented on quantum/split_common/serial.c:
```diff
@@ -29,36 +29,34 @@
   #endif
 
+  #define setPinInputHigh(pin) do {\
+          DDRx_ADDRESS(pin)  &= ~ _BV(pin & 0xF);\
```
```suggestion
          DDRx_ADDRESS(pin)  &= ~_BV(pin & 0xF);\
```

mtei replied:
Thanks.

mtei commented:
I deliberately avoided serial.c to depend on quantum.h. serial.c is a very timing-dependent module, so we want to avoid other impacts as much as possible.

drashna commented:
In that case, would it make sense to move both out of the respective files, and into a separate header file, where both `quantum.h` and `serial.c` could include them?   

That way, we're not maintaining multiple copies of the GPIO commands, and don't have to worry about keeping them in sync (for behavioral reasons)?

mtei commented:
> In that case, would it make sense to move both out of the respective files, and into a separate header file, where both `quantum.h` and `serial.c` could include them?

I already thought about that method, but I did not adopt it. If serial.c imports gpio_control.h, you need to be careful modifying gpio_control.h as you would for serial.c. It's not desirable.

> That way, we're not maintaining multiple copies of the GPIO commands, and don't have to worry about keeping them in sync (for behavioral reasons)?

Do you need to sync? I do not think that it is necessary.  It is dangerous to change the working serial.c just to synchronize.

fauxpark commented:
> If serial.c imports gpio_control.h, you need to be careful modifying gpio_control.h as you would for serial.c.

Not sure I understand. Of course you need to be careful modifying these GPIO macros, however they are fairly simple so I would be surprised if there were any major problems with them that haven't been discovered yet. If something *does* need to be changed, they must *both* be changed anyway, because otherwise you have two conflicting `#defines` which will probably make GCC complain.

The thing is that serial.c shouldn't need to know nor define how eg. `setPinOutput()` works at this level. All of that should be abstracted away.

mtei commented:
In general I like to be abstract. However, the abstraction is not necessary only for serial.c. What is important is that the generated binaries do not change.

We have already experienced that when the GCC version changes the generation code will change and the timing changes. (#4269)  We can't trust the simplicity of the source code's appearance.

drashna commented:
It looks like there is a merge conflict here, now.

mtei commented:
Thanks, I fixed.

Then I checked again that there was no change in the binary generated by this PR.
```
$ git checkout avr_gpio_macro_more_readable
Switched to branch 'avr_gpio_macro_more_readable'
$ make SKIP_DEBUG_INFO=y SKIP_VERSION=y lets_split:default:clean
$ make SKIP_DEBUG_INFO=y SKIP_VERSION=y lets_split:default
....
 * The firmware size is fine - 19450/28672 (67%, 9222 bytes free)
$ mv lets_split_rev2_default.hex new-lets_split_rev2_default.hex

$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
$ make SKIP_DEBUG_INFO=y SKIP_VERSION=y lets_split:default:clean
$ make SKIP_DEBUG_INFO=y SKIP_VERSION=y lets_split:default
....
 * The firmware size is fine - 19450/28672 (67%, 9222 bytes free)

$ md5 *lets_split_rev2_default.hex
MD5 (lets_split_rev2_default.hex) = 4999375ced50a8e7f731bc06bdeb534f
MD5 (new-lets_split_rev2_default.hex) = 4999375ced50a8e7f731bc06bdeb534f

```
